### PR TITLE
TASK: Replace PHPCS with PHP-CS-Fixer to enforce coding guidelines

### LINKED ISSUE #4580: Adjust linter to report unused namespace imports
There are quite a lot of unused `use` statements in our code base.

It would be helpful if `composer lint` would report these.

~There should be a `NoUsedImports` feature in PHPCS, but I can't get it to work somehow~

There is no such sniff in PHPCS according to https://github.com/squizlabs/PHP_CodeSniffer/issues/801 but we could use the php-cs-fixer for that:
```
php-cs-fixer fix --rules no_unused_imports ...
```

### PR DESCRIPTION
**Pull request by bwaidelich**
Adds [PHP-CS-Fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer) as dev dependency as replacement for [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer) and configures the following rules:
* `PSR12` [PSR-12 Coding Guideline](https://www.php-fig.org/psr/psr-12/) – that was previously covered by PHPCS
* `no_unused_imports` disallows unused PHP namespace imports
* `ordered_imports` requires `use` statements to be ordered alphabetically

It makes those checks part of our `composer:lint` script and adds:

* `composer lint:php-cs-fixer` to check them and
* `composer lint:php-cs-fixer:fix` to fix violations
* `composer lint:fix` to fix any linter violations (currently the same)

Resolves: #4580

### CODE REVIEW

**Comment by bwaidelich** [comment]
This build will fail.
Locally I get:
> Found 237 of 739 files that can be fixed in 4.083 seconds, 22.000 MB memory used

When I run `composer lint:php-cs-fixer:fix` it fixes those namespace imports and afterwards I get
> Found 0 of 739 files that can be fixed in 0.030 seconds, 18.000 MB memory used

**Comment by mhsdesign** [comment]
Its a bit sad that we really need another dev dependency for this :/

That makes stuff just needless complex 😅 

i currently only use `phpstan` locally to be honest and even ignore `phpcs` as it slows me down :P

**Comment by bwaidelich** [comment]
> Its a bit sad that we really need another dev dependency for this :/
> That makes stuff just needless complex 

It's just a dev dependency, it doesn't affect complexity of our code and in the mid-term can even help us to keep complexity lower.
But of course there is a cost attached to any new dependency..

> i currently only use phpstan locally to be honest and even ignore phpcs as it slows me down :P

You are free to ignore it, but our CI won't and IMO that's a good thing that helps us to establish a common coding style.
PHPStorm has built-in support for all three quality tools that highlight errors right away. And if configured correctly, it will create code in the same style and fix it.

Long story short: I think it's valuable to have the namespace imports cleaned up by a tool (and it was _you_ that brought this up *g). PHPCS won't be able to do this, and from reading the comments on https://github.com/squizlabs/PHP_CodeSniffer/pull/1106 it will most likely never support it.

Having said this, maybe PHP-CS-Fixer supports the other checks that we currently do with PHPCS, so we could at least _replace_ it!?

**Comment by bwaidelich** [comment]
> Having said this, maybe PHP-CS-Fixer supports the other checks that we currently do with PHPCS, so we could at least replace it!?

That's seems to be the case..
@mhsdesign What do you think of b418924dea957076da84fd893c022a55ec59ae9b ?

*Note:* This is just an experiment. We should all agree if whether want to switch the quality tools

**Comment by mhsdesign** [comment]
Thanks a lot @bwaidelich !

I must say this looks good in my eyes, but i have only used phpstan so far and let php cs be handled by the ci (rather i forget about its existence all the time).

Getting rid of php cs might be a good idea as they have a super slow release cycle and no clear maintainer, which lead to weird hotfixes and lost time in the past.

So overall im not opinionated in which tool to learn / use but do plead for simplicity so 1 tool should be enough 😅

**Comment by mhsdesign** [comment]
A super great goal would be imo to then get also rid of our `styleci`, as i havent figured out how to run it locally and this is currently always a real bummer when developing as i always forget how many spaces before an anonymous function :D.

So if we can reduce the complexity from two linters to only php cs fixer that would great and im happy to assist.

**Comment by kitsunet** [comment]
I would like this too and kill styleci

**Comment by mficzel** [comment]
Also agree that this is a huge benefit compared to styleci as anyone can execute it locally.
Way too many prs take an additional test cycles with waiting time because styleci decides to suddenly not like something.

**Review by mficzel — APPROVED** [review]
Fine by :eyes:

**Comment by mhsdesign** [comment]
@mficzel this change needs some new love and needs to be rebased as its from 2024 feel free to take over id say

**Comment by mhsdesign** [comment]
superseeded by https://github.com/neos/neos-development-collection/pull/5928
